torch/csrc/api/include/torch/optim/serialize.h

```diff
@@ -207,27 +207,77 @@ void serialize(
   }
 }
 
+template <typename T>
+c10::List<T> deque_to_list(const std::deque<T>& dq) {
+  c10::List<Tensor> list;
+  list.reserve(dq.size());
+  for (const auto& e : dq) {
+    list.push_back(e);
+  }
+  return list;
+}
+
+template <typename T>
+void list_to_deque(const c10::List<T>& list, std::deque<T>& dq) {
+  for (size_t i = 0; i < list.size(); i++) {
+    dq.push_back(list[i]);
+  }
+}
+
 #define _TORCH_OPTIM_SERIALIZE(name) \
   torch::optim::serialize(archive, #name, self.name)
 
 #define _TORCH_OPTIM_SERIALIZE_WITH_TEMPLATE_ARG(OptimizerName) \
   torch::optim::serialize<OptimizerName##ParamState, OptimizerName##Options>(archive, self)
 
-#define _TORCH_OPTIM_SERIALIZE_TORCH_ARG(name) \
-{ \
+#define _TORCH_OPTIM_SERIALIZE_TORCH_ARG(name) { \
   auto ivalue = torch::IValue(name()); \
   /* do not serialize if name is an undefined tensor*/ \
   if (!(ivalue.isTensor() && ivalue.unsafeToTensorImpl() == at::UndefinedTensorImpl::singleton())) { \
     archive.write(#name, ivalue); \
   } \
 }
 
+#define _TORCH_OPTIM_SERIALIZE_TORCH_ARG_DEQUE(name) {\
+  if (name().size() != 0) { \
+    c10::IValue ivalue = torch::IValue(deque_to_list(name())); \
+    archive.write(#name, ivalue); \
+  } \
+}
+
 #define _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(T, name) \
 { \
   c10::IValue ivalue; \
   bool exists = archive.try_read(#name, ivalue); \
   if (exists) \
     name(ivalue.to<T>()); \
+  /* undefined tensors are not serialized. it's imp to save {} in name() to ensure that
+  undefined tensor values are retained in case the user set it. */ \
+  else if (ivalue.isTensor()) \
```

Comment:
When we hit this second branch, I don't think `ivalue.isTensor()` would ever be true, because the value simply doesn't exist in the archive and there is no way to infer what the tag of the `ivalue` is.

Reply:
I wonder if it would make sense to skip all this change to the `_TORCH_OPTIM_DESERIALIZE_TORCH_ARG` macro, based on two facts:
1. Optimizer options don't take tensor as argument, and thus there is no need to preserve user-specified undefined tensor value
2. Optimizer param state does have tensor as argument. However, in Python API we don't serialize None tensor in param state, and in C++ API we don't need to do it either

Reply:
yeah that makes sense. I agree with that choice

Reply:
It would be great to put this comment explicitly in the code, maybe above `TORCH_API serialize::OutputArchive& operator<<(` in `torch/csrc/api/include/torch/optim/optimizer.h`:



> How do we decide whether to serialize undefined tensors or `c10::nullopt` values into the output archive?
> Answer: we strictly follow the behavior of Python API. To be more specific:
> 1. For optimizer options:
>   a) For undefined tensor: currently no tensor is used as an options argument in Python API, so we don't need to worry about it now.
>   b) For `c10::nullopt` value: we serialize `c10::nullopt` values into the output archive, to follow the exact same behavior as Python API.
> 2. For optimizer param state:
>   a) For undefined tensor: in param state, undefined tensor in C++ impl is equivalent to missing key in Python impl. Since we don't serialize missing keys in Python API, we skip undefined tensors when serializing the param state.
>   b) For `c10::nullopt` value: in param state, `c10::nullopt` value in C++ impl is equivalent to missing key in Python impl. Since we don't serialize missing keys in Python API, we skip `c10::nullopt` values when serializing the param state.